utils.groovy

```diff
@@ -15,19 +25,51 @@ def load_pipecfg() {
     def url = jenkinscfg['pipecfg-url']
     def ref = jenkinscfg['pipecfg-ref']
 
+    def pipecfg
     if (url == 'in-tree') {
-        return readYaml(file: "config.yaml")
+        pipecfg = readYaml(file: "config.yaml")
+    } else {
+        // this uses the `checkout` workflow step instead of just manually cloning so
+        // that changes show up in the Jenkins UI
+        checkout([
+            $class: 'GitSCM',
+            branches: [[name: "origin/${ref}"]],
+            extensions: [[$class: 'RelativeTargetDirectory', relativeTargetDir: 'pipecfg']],
+            userRemoteConfigs: [[url: url]]
+        ])
+        pipecfg = readYaml(file: "pipecfg/config.yaml")
     }
+    validate_pipecfg(pipecfg)
+    return pipecfg
+}
 
-    // this uses the `checkout` workflow step instead of just manually cloning so
-    // that changes show up in the Jenkins UI
-    checkout([
-        $class: 'GitSCM',
-        branches: [[name: "origin/${ref}"]],
-        extensions: [[$class: 'RelativeTargetDirectory', relativeTargetDir: 'pipecfg']],
-        userRemoteConfigs: [[url: url]]
-    ])
-    return readYaml(file: "pipecfg/config.yaml")
+def validate_pipecfg(pipecfg) {
+    if (env.JENKINS_URL in PROTECTED_JENKINSES) {
+        // We're running on a protected Jenkins instance. Check the hotfix
+        // build policy and that S3 buckets are as expected.
+        def (bucket, key, allow_hotfixes) = PROTECTED_JENKINSES[env.JENKINS_URL]
+        if (pipecfg.hotfix && !allow_hotfixes) {
+            error("Hotfix builds are not allowed on ${env.JENKINS_URL}")
+        }
+        if (pipecfg.s3 && (bucket != pipecfg.s3.bucket || key != pipecfg.s3.builds_key)) {
+            def target_bucket_key = "${pipecfg.s3.bucket}/${pipecfg.s3.builds_key}"
+            error("S3 bucket/key on ${env.JENKINS_URL} must be ${bucket}/${key} (found ${target_bucket_key})")
+        }
+    } else if (AWSBuildUploadCredentialExists()) {
+        // We're *not* running on a protected Jenkins instance and we have AWS
+        // creds available. Check that we're not trying to push to a protected
+        // S3 bucket.
+
+        // Transform PROTECTED_JENKINSES to a map of {bucket -> jenkins URL}
+        // for easier lookup and reporting
+        def protected_buckets = PROTECTED_JENKINSES.collectEntries{jenkins_url, v ->
+            def (bucket, key, allow_hotfixes) = v
+            [bucket, jenkins_url]
+        }
+        if (pipecfg.s3.bucket in protected_buckets) {
+            error("S3 bucket ${pipecfg.s3.bucket} can only be used on ${protected_buckets[pipecfg.s3.bucket]}")
+        }
```

Comment:
So here we are saying that no one can use the same bucket as a not-hotfix build on these jenkins instances. I imagine they might argue that they want to use the same bucket but a different path within the bucket, which we don't allow for here.

Reply:
Almost. This is in the non-protected Jenkins instance branch. We're checking that such Jenkins instances don't try to upload artifacts to the canonical S3 buckets.

The hotfix work in the next PR does support writing to the same bucket but different path for hotfix builds. The expectation is that they'll be using the same Jenkins instance for regular and hotfix builds.

Reply:
oh OK. +1